### PR TITLE
Show asterisk in flyover text if question text is not present and questionnaire item is required.

### LINKED ISSUE #544: Add "required" text for required fields in Questionaire
**Is your feature request related to a problem? Please describe.**
No this is just an enhancement.

**Describe the solution you'd like**
Asterisks symbol can be added for the fields which have an attribute ``"required": "true"``.
``*`` indicates required fields as seen in the sample below.

Screenshot 2021-06-11 002613

**Describe alternatives you've considered**
NA

**Additional context**
NA.

**Would you like to work on the issue?**
Yes

### PR DESCRIPTION
**Pull request by santosh-pingle**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #544

**Description**
Clear and concise code change description. 

- If question text is not present and questionnaire item is required then show asterisk in flyover text of sdc widgets which uses text input text layout (Edit text).

 

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: Feature 

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

**Comment by santosh-pingle** [comment]
@shelaghm 
All attached screenshots does not have question text and questionnaire item is required (mandatory).

Screen Shot 2022-12-12 at 3 41 54 PM
Screen Shot 2022-12-12 at 3 42 26 PM
Screen Shot 2022-12-12 at 3 42 57 PM
Screen Shot 2022-12-12 at 3 43 33 PM
Screen Shot 2022-12-12 at 3 44 04 PM

**Comment by omarismail94** [comment]
@santosh-pingle did we resolve whether we wanted `* required`  to appear on the bottom of all questions?

**Comment by santosh-pingle** [comment]
> @santosh-pingle did we resolve whether we wanted `* required` to appear on the bottom of all questions?

I will start this thread in issue #544.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/MoreQuestionnaireItemView.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
+/*
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/MoreQuestionnaireItemView.kt]: done.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/MoreQuestionnaireItemView.kt [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import com.google.android.fhir.datacapture.localizedTextSpanned
+import com.google.android.material.textfield.TextInputLayout
+import org.hl7.fhir.r4.model.Questionnaire
+
+/**
+ * Appends asterisk to the flyover text if question text
+ * [Questionnaire.QuestionnaireItemComponent.localizedTextSpanned] is null or blank and
+ * questionnaire item is must be answered.
+ */
+inline fun TextInputLayout.showAsteriskInFlyoverText(
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/MoreQuestionnaireItemView.kt]: it was typo mistake, i meant internal and not inline.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/MoreQuestionnaireItemView.kt [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  questionnaireItem: Questionnaire.QuestionnaireItemComponent
+) {
+  if (!questionnaireItem.required || !questionnaireItem.localizedTextSpanned.isNullOrBlank()) {
+    return
+  }
+  showAsteriskInFlyoverText(questionnaireItem.localizedFlyoverSpanned.toString())
+}
+
+/** Appends asterisk to the argument [flyoverText]. */
+inline fun TextInputLayout.showAsteriskInFlyoverText(flyoverText: String) {
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/MoreQuestionnaireItemView.kt]: done.

**Comment by omarismail94** [comment]
Thanks @santosh-pingle ! LGTM!

**Comment by omarismail94** [comment]
We will wait for @shelaghm actually to get back to @jingtang10 's comment in the Issue #544  before approving

**Comment by shelaghm** [comment]
@omarismail94 @santosh-pingle I've commented over on https://github.com/google/android-fhir/issues/544 let me know if either of you have thoughts as well :)

**Review by jingtang10 — CHANGES_REQUESTED** [review]
i think we're gonna not use asterisk at all as the result of UX research findings and discussions with @shelaghm.

sorry that you'll need to rework this pr.

**Comment by shelaghm** [comment]
Let's wait while we get some input from the community on how they use required and then make a decision on how we re-work this issue. So far we've heard from Ona's designers. Will be great to hear from a few others how they are using this feature in their deployments.

**Comment by shelaghm** [comment]
FYI https://github.com/google/android-fhir/issues/544 Ona uses `Optional` instead of `Required`.

**Comment by FikriMilano** [comment]
@santosh-pingle can we make this configurable? would like to show nothing if required and show `Optional` if not required

**Comment by jingtang10** [comment]
@santosh-pingle can we close this PR now given it's also out of date?

we can reopen/create a new PR once we have conclusion on UX.

**Comment by jingtang10** [comment]
thanks!

**Pull request closed without merging**